Sources/WebSocket/WebSocket.swift

```diff
@@ -91,147 +98,365 @@ public final class WebSocket: WebSocketProtocol {
+    func host(with urlComponents: URLComponents) throws -> NWEndpoint.Host {
+        guard let host = urlComponents.host else {
+            throw WebSocketError.invalidURLComponents(urlComponents)
         }
+        return NWEndpoint.Host(host)
+    }
+    
+    func port(with urlComponents: URLComponents) throws -> NWEndpoint.Port {
+        if let raw = urlComponents.port, let port = NWEndpoint.Port(rawValue: UInt16(raw)) {
+            return port
+        } else if "ws" == urlComponents.scheme {
+             return NWEndpoint.Port.http
+         } else if "wss" == urlComponents.scheme {
+             return NWEndpoint.Port.https
+         } else  {
+             throw WebSocketError.invalidURLComponents(urlComponents)
+         }
+    }
+    
+    func parameters(with urlComponents: URLComponents) throws -> NWParameters {
+        let parameters: NWParameters
+        switch urlComponents.scheme {
+        case "ws":
+            parameters = .tcp
+        case "wss":
+            parameters = .tls
+        default:
+            throw WebSocketError.invalidURLComponents(urlComponents)
+        }
+        
+        let webSocketOptions = NWProtocolWebSocket.Options()
+        parameters.defaultProtocolStack.applicationProtocols.insert(webSocketOptions, at: 0)
+
+        return parameters
     }
 }
 
-// MARK: URLSessionWebSocketDelegate
+private typealias OnStateUpdateHandler = (NWConnection.State) -> Void
+private typealias OnReceiveMessageHandler = (Data?, NWConnection.ContentContext?, Bool, NWError?) -> Void
 
-private class WebSocketDelegate: NSObject, URLSessionWebSocketDelegate {
-    private let onOpen: OnOpenHandler
-    private let onClose: OnCloseHandler
-    private let onCompletion: OnCompletionHandler
+// MARK: onOpen and onClose
 
-    init(onOpen: @escaping OnOpenHandler,
-         onClose: @escaping OnCloseHandler,
-         onCompletion: @escaping OnCompletionHandler)
-    {
-        self.onOpen = onOpen
-        self.onClose = onClose
-        self.onCompletion = onCompletion
-        super.init()
-    }
+private extension WebSocket  {
+    var onStateUpdate: OnStateUpdateHandler {
+        return { [weak self] (state) -> Void in
+            guard let self = self else { return }
 
-    func urlSession(_ webSocketSession: URLSession,
-                    webSocketTask: URLSessionWebSocketTask,
-                    didOpenWithProtocol protocol: String?) {
-        self.onOpen(webSocketSession, webSocketTask, `protocol`)
-    }
+            Swift.print(#function, String(describing: state))
 
-    func urlSession(_ session: URLSession,
-                    webSocketTask: URLSessionWebSocketTask,
-                    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
-                    reason: Data?) {
-        self.onClose(session, webSocketTask, closeCode, reason)
+            switch state {
+            case .setup:
+                break
+            case .waiting:
+                break
```

Comment:
Is there anything interesting to note if we are `.waiting`?

Reply:
Yes, there is. Waiting _(sometimes, usually, always?)_ indicates the device doesn't have an Internet connection. Apple has been very, very strongly discouraging the user of Reachability in apps now and, instead, offers ways like this to know when and why the device isn't connected to the Internet.